QP: fix panic on an unknown field in referenced table

## Steps to reproduce the problem

```sql
TABLE cdoc1 INHERITS CDoc ();
TABLE cdoc2 INHERITS CDoc (
	field2 ref(cdoc1),
);
```
1. insert cdoc1
2. insert cdoc2: field2 = inserted cdoc1 ID
3. q.sys.Collection where let's ask to get an unknown field from cdoc1 taken by ref from codc2.field2, i.e.:
```
{"args":{"Schema":"app1pkg.cdoc2"},"elements":[{"fields": ["field2","sys.ID"], "refs":[["field2","unexistingFieldInTargetDoc"]]}]}
```

### Result
```
panic: unsupported kind DataKind_null for field unexistingFieldInTargetDoc [recovered]
	panic: unsupported kind DataKind_null for field unexistingFieldInTargetDoc

goroutine 3362 [running]:
github.com/voedger/voedger/pkg/coreutils.ReadByKind.func1()
	c:/workspace/go/voedger/pkg/coreutils/objectreader.go:21 +0x218
panic({0x12da2c0?, 0xc005eeb930?})
	C:/Program Files/Go/src/runtime/panic.go:785 +0x132
github.com/voedger/voedger/pkg/coreutils.ReadByKind({0xc005e9fac0, 0x1a}, 0xc0?, {0x15e14b0?, 0xc005ef3dd0?})
	c:/workspace/go/voedger/pkg/coreutils/objectreader.go:44 +0x419
github.com/voedger/voedger/pkg/processors/query.(*EnrichmentOperator).DoAsync(0xc005ee9d00, {0x15da460, 0xc005eb1590}, {0x15d39e0, 0xc005ee9dc0})
	c:/workspace/go/voedger/pkg/processors/query/operator-enrichment-impl.go:57 +0x54a
github.com/voedger/voedger/pkg/pipeline.(*WiredOperator).doAsync(0xc0068c03f0, {0x15d39e0, 0xc005ee9dc0})
	c:/workspace/go/voedger/pkg/pipeline/wired-operator.go:90 +0xec
github.com/voedger/voedger/pkg/pipeline.puller_async(0xc0068c03f0)
	c:/workspace/go/voedger/pkg/pipeline/async.go:36 +0x305
created by github.com/voedger/voedger/pkg/pipeline.NewAsyncPipeline in goroutine 70
	c:/workspace/go/voedger/pkg/pipeline/async-pipeline-impl.go:54 +0x82f
```
server restarts

## Expected behavior
error that says that `unexistingFieldInTargetDoc` field is missing in the referenced `cdoc1`

## Technical solution
- wrong field -> return an appropriate error in `EnrichmentOperator`
- QP: error in `rowsProcessor` async pipeline -> send it to QP, do not close `IResultSenderCloseable` with that error. QP must close it itself with errors from query exec or from `rowsProcessor`

--- a/pkg/processors/query/impl.go
+++ b/pkg/processors/query/impl.go
@@ -39,7 +39,7 @@ import (
 )
 
 func implRowsProcessorFactory(ctx context.Context, appDef appdef.IAppDef, state istructs.IState, params IQueryParams,
-	resultMeta appdef.IType, rs IResultSenderClosable, metrics IMetrics) pipeline.IAsyncPipeline {
+	resultMeta appdef.IType, rs IResultSenderClosable, metrics IMetrics, errCh chan<- error) pipeline.IAsyncPipeline {
 	operators := make([]*pipeline.WiredOperator, 0)
 	if resultMeta == nil {
 		// happens when the query has no result, e.g. q.air.UpdateSubscriptionDetails
@@ -86,6 +86,7 @@ func implRowsProcessorFactory(ctx context.Context, appDef appdef.IAppDef, state 
 	operators = append(operators, pipeline.WireAsyncOperator("Send to bus", &SendToBusOperator{
 		rs:      rs,
 		metrics: metrics,
+		errCh:   errCh,
 	}))
 	return pipeline.NewAsyncPipeline(ctx, "Rows processor", operators[0], operators[1:]...)
 }
@@ -129,6 +130,13 @@ func implServiceFactory(serviceChannel iprocbus.ServiceChannel, resultSenderClos
 					if qwork.rowsProcessor != nil {
 						// wait until all rows are sent
 						qwork.rowsProcessor.Close()
+					}
+					select {
+					case rowsProcErr := <-qwork.rowsProcessorErrCh:
+						if err == nil {
+							err = rowsProcErr
+						}
+					default:
 					}
 					err = coreutils.WrapSysError(err, http.StatusInternalServerError)
 					rs.Close(err)
@@ -358,7 +366,7 @@ func newQueryProcessorPipeline(requestCtx context.Context, authn iauthnz.IAuthen
 				qw.metrics.Increase(buildSeconds, time.Since(now).Seconds())
 			}()
 			qw.rowsProcessor = ProvideRowsProcessorFactory()(qw.msg.RequestCtx(), qw.appStructs.AppDef(),
-				qw.state, qw.queryParams, qw.resultType, qw.rs, qw.metrics)
+				qw.state, qw.queryParams, qw.resultType, qw.rs, qw.metrics, qw.rowsProcessorErrCh)
 			return nil
 		}),
 	}
@@ -371,22 +379,23 @@ type queryWork struct {
 	rs       IResultSenderClosable
 	appParts appparts.IAppPartitions
 	// work
-	requestData       map[string]interface{}
-	state             state.IHostState
-	queryParams       IQueryParams
-	appPart           appparts.IAppPartition
-	appStructs        istructs.IAppStructs
-	resultType        appdef.IType
-	execQueryArgs     istructs.ExecQueryArgs
-	maxPrepareQueries int
-	rowsProcessor     pipeline.IAsyncPipeline
-	metrics           IMetrics
-	principals        []iauthnz.Principal
-	principalPayload  payloads.PrincipalPayload
-	secretReader      isecrets.ISecretReader
-	iWorkspace        appdef.IWorkspace
-	iQuery            appdef.IQuery
-	wsDesc            istructs.IRecord
+	requestData        map[string]interface{}
+	state              state.IHostState
+	queryParams        IQueryParams
+	appPart            appparts.IAppPartition
+	appStructs         istructs.IAppStructs
+	resultType         appdef.IType
+	execQueryArgs      istructs.ExecQueryArgs
+	maxPrepareQueries  int
+	rowsProcessor      pipeline.IAsyncPipeline
+	rowsProcessorErrCh chan error // will contain the first error from rowProcessor if any. The rest of errors in rowsProcessor will be just logged
+	metrics            IMetrics
+	principals         []iauthnz.Principal
+	principalPayload   payloads.PrincipalPayload
+	secretReader       isecrets.ISecretReader
+	iWorkspace         appdef.IWorkspace
+	iQuery             appdef.IQuery
+	wsDesc             istructs.IRecord
 	// queryExec         func(ctx context.Context, args istructs.ExecQueryArgs, callback istructs.ExecQueryCallback) error
 	callbackFunc istructs.ExecQueryCallback
 }
@@ -394,13 +403,14 @@ type queryWork struct {
 func newQueryWork(msg IQueryMessage, rs IResultSenderClosable, appParts appparts.IAppPartitions,
 	maxPrepareQueries int, metrics *queryProcessorMetrics, secretReader isecrets.ISecretReader) *queryWork {
 	return &queryWork{
-		msg:               msg,
-		rs:                rs,
-		appParts:          appParts,
-		requestData:       make(map[string]interface{}),
-		maxPrepareQueries: maxPrepareQueries,
-		metrics:           metrics,
-		secretReader:      secretReader,
+		msg:                msg,
+		rs:                 rs,
+		appParts:           appParts,
+		requestData:        make(map[string]interface{}),
+		maxPrepareQueries:  maxPrepareQueries,
+		metrics:            metrics,
+		secretReader:       secretReader,
+		rowsProcessorErrCh: make(chan error, 1),
 	}
 }
 

--- a/pkg/processors/query/impl_test.go
+++ b/pkg/processors/query/impl_test.go
@@ -147,7 +147,8 @@ func TestBasicUsage_RowsProcessorFactory(t *testing.T) {
 		},
 		close: func(err error) {},
 	}
-	processor := ProvideRowsProcessorFactory()(context.Background(), appDef, s, params, resultMeta, rs, &testMetrics{})
+	rowsProcessorErrCh := make(chan error, 1)
+	processor := ProvideRowsProcessorFactory()(context.Background(), appDef, s, params, resultMeta, rs, &testMetrics{}, rowsProcessorErrCh)
 
 	require.NoError(processor.SendAsync(work(1, "Cola", 10)))
 	require.NoError(processor.SendAsync(work(3, "White wine", 20)))
@@ -156,6 +157,11 @@ func TestBasicUsage_RowsProcessorFactory(t *testing.T) {
 	processor.Close()
 
 	require.Equal(`[[[3,"White wine","Alcohol drinks"]]]`, result)
+	select {
+	case err := <-rowsProcessorErrCh:
+		t.Fatal(err)
+	default:
+	}
 }
 
 func deployTestAppWithSecretToken(require *require.Assertions,
@@ -442,7 +448,8 @@ func TestRawMode(t *testing.T) {
 		},
 		close: func(err error) {},
 	}
-	processor := ProvideRowsProcessorFactory()(context.Background(), appDef, &mockState{}, queryParams{}, resultMeta, rs, &testMetrics{})
+	rowsProcessorErrCh := make(chan error, 1)
+	processor := ProvideRowsProcessorFactory()(context.Background(), appDef, &mockState{}, queryParams{}, resultMeta, rs, &testMetrics{}, rowsProcessorErrCh)
 
 	require.NoError(processor.SendAsync(rowsWorkpiece{
 		object: &coreutils.TestObject{
@@ -456,6 +463,11 @@ func TestRawMode(t *testing.T) {
 		},
 	}))
 	processor.Close()
+	select {
+	case err := <-rowsProcessorErrCh:
+		t.Fatal(err)
+	default:
+	}
 
 	require.Equal(`[[["[accepted]"]]]`, result)
 }

--- a/pkg/processors/query/interface.go
+++ b/pkg/processors/query/interface.go
@@ -26,7 +26,7 @@ import (
 // The pipeline is used to process data fetched by QueryHandler
 // TODO In my opinion we have to remove it from export
 type RowsProcessorFactory func(ctx context.Context, appDef appdef.IAppDef, state istructs.IState,
-	params IQueryParams, resultMeta appdef.IType, rs IResultSenderClosable, metrics IMetrics) pipeline.IAsyncPipeline
+	params IQueryParams, resultMeta appdef.IType, rs IResultSenderClosable, metrics IMetrics, errCh chan<- error) pipeline.IAsyncPipeline
 
 type ServiceFactory func(serviceChannel iprocbus.ServiceChannel, resultSenderClosableFactory ResultSenderClosableFactory,
 	appParts appparts.IAppPartitions, maxPrepareQueries int, metrics imetrics.IMetrics, vvm string,

--- a/pkg/processors/query/operator-enrichment-impl.go
+++ b/pkg/processors/query/operator-enrichment-impl.go
@@ -6,6 +6,8 @@ package queryprocessor
 
 import (
 	"context"
+	"fmt"
+	"net/http"
 	"time"
 
 	"github.com/voedger/voedger/pkg/appdef"
@@ -50,7 +52,11 @@ func (o *EnrichmentOperator) DoAsync(ctx context.Context, work pipeline.IWorkpie
 				record := sv.AsRecord("")
 
 				recFields := o.fieldsDefs.get(record.QName())
-				value := coreutils.ReadByKind(field.RefField(), recFields[field.RefField()], record)
+				refFieldKind, ok := recFields[field.RefField()]
+				if !ok {
+					return work, coreutils.NewHTTPErrorf(http.StatusBadRequest, fmt.Errorf("ref field %s references to table %s that does not contain field %s", field.Field(), record.QName(), field.RefField()))
+				}
+				value := coreutils.ReadByKind(field.RefField(), refFieldKind, record)
 				if element.Path().IsRoot() {
 					work.(IWorkpiece).PutEnrichedRootFieldKind(field.Key(), recFields[field.RefField()])
 				}

--- a/pkg/processors/query/operator-send-to-bus-impl.go
+++ b/pkg/processors/query/operator-send-to-bus-impl.go
@@ -8,6 +8,7 @@ import (
 	"context"
 	"time"
 
+	"github.com/voedger/voedger/pkg/goutils/logger"
 	"github.com/voedger/voedger/pkg/pipeline"
 )
 
@@ -16,6 +17,7 @@ type SendToBusOperator struct {
 	rs          IResultSenderClosable
 	initialized bool
 	metrics     IMetrics
+	errCh       chan<- error
 }
 
 func (o *SendToBusOperator) DoAsync(_ context.Context, work pipeline.IWorkpiece) (outWork pipeline.IWorkpiece, err error) {
@@ -32,5 +34,9 @@ func (o *SendToBusOperator) DoAsync(_ context.Context, work pipeline.IWorkpiece)
 }
 
 func (o *SendToBusOperator) OnError(_ context.Context, err error) {
-	o.rs.Close(err)
+	select {
+	case o.errCh <- err:
+	default:
+		logger.Error("failed to send error from rowsProcessor to QP: " + err.Error())
+	}
 }

--- a/pkg/processors/query/operator-send-to-bus-impl_test.go
+++ b/pkg/processors/query/operator-send-to-bus-impl_test.go
@@ -6,6 +6,7 @@ package queryprocessor
 
 import (
 	"context"
+	"errors"
 	"testing"
 
 	"github.com/stretchr/testify/require"
@@ -14,6 +15,7 @@ import (
 func TestSendToBusOperator_DoAsync(t *testing.T) {
 	require := require.New(t)
 	counter := 0
+	errCh := make(chan error, 1)
 	operator := SendToBusOperator{
 		rs: testResultSenderClosable{
 			sendElement: func(name string, element interface{}) (err error) {
@@ -25,6 +27,7 @@ func TestSendToBusOperator_DoAsync(t *testing.T) {
 			},
 		},
 		metrics: &testMetrics{},
+		errCh:   errCh,
 	}
 	work := rowsWorkpiece{outputRow: &testOutputRow{values: []interface{}{"hello world"}}}
 
@@ -34,4 +37,29 @@ func TestSendToBusOperator_DoAsync(t *testing.T) {
 	require.Equal(work, outWork)
 	require.NoError(err)
 	require.Equal(1, counter)
+}
+
+func TestSendToBusOperator_OnError(t *testing.T) {
+	require := require.New(t)
+	errCh := make(chan error, 1)
+	testError := errors.New("test error")
+	operator := SendToBusOperator{
+		rs: testResultSenderClosable{
+			sendElement: func(name string, element interface{}) (err error) {
+				return testError
+			},
+			startArraySection: func(sectionType string, path []string) {},
+		},
+		metrics: &testMetrics{},
+		errCh:   errCh,
+	}
+
+	operator.OnError(context.Background(), testError)
+
+	select {
+	case err := <-errCh:
+		require.ErrorIs(testError, err)
+	default:
+		t.Fail()
+	}
 }

--- a/pkg/sys/it/bug_test.go
+++ b/pkg/sys/it/bug_test.go
@@ -7,6 +7,7 @@ package sys_it
 import (
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"net/http"
 	"testing"
@@ -105,4 +106,21 @@ func Test409OnRepeatedlyUsedRawIDsInResultCUDs_(t *testing.T) {
 	}
 	ws := vit.WS(istructs.AppQName_test1_app1, "test_ws")
 	vit.PostWS(ws, "c.app1pkg.MockCmd", `{"args":{"Input":"Str"}}`, coreutils.Expect409()).Println()
+}
+
+// https://github.com/voedger/voedger/issues/2759
+func TestWrongFieldReferencedByRefField(t *testing.T) {
+	vit := it.NewVIT(t, &it.SharedConfig_App1)
+	defer vit.TearDown()
+
+	ws := vit.WS(istructs.AppQName_test1_app1, "test_ws")
+
+	body := `{"cuds":[{"fields":{"sys.ID": 1,"sys.QName":"app1pkg.cdoc1"}}]}`
+	cdoc1ID := vit.PostWS(ws, "c.sys.CUD", body).NewID()
+
+	body = fmt.Sprintf(`{"cuds":[{"fields":{"sys.ID": 1,"sys.QName":"app1pkg.cdoc2","field2": %d}}]}`, cdoc1ID)
+	vit.PostWS(ws, "c.sys.CUD", body).NewID()
+
+	body = `{"args":{"Schema":"app1pkg.cdoc2"},"elements":[{"fields": ["field2","sys.ID"], "refs":[["field2","unexistingFieldInTargetDoc"]]}]}`
+	vit.PostWS(ws, "q.sys.Collection", body, coreutils.Expect400("ref field field2 references to table app1pkg.cdoc1 that does not contain field unexistingFieldInTargetDoc"))
 }

